component/azstorage/block_blob.go

```diff
@@ -1068,119 +1044,246 @@ func (bb *BlockBlob) TruncateFile(name string, size int64) error {
+		err = bb.createSparseFile(name, size, blockSize, make([]string, 0), 16)
 		if err != nil {
-			log.Err("BlockBlob::TruncateFile : Failed to read small file %s", name, err.Error())
-			return err
+			log.Err("BlockBlob::TruncateFile : Failed to expand file %s [%s]", name, err.Error())
 		}
-		err = bb.WriteFromBuffer(name, nil, data)
+		return err
+	}
+
+	// Get the committed block list from the source file
+	committedBlockList, err := bb.GetCommittedBlockList(name)
+	if err != nil {
+		log.Err("BlockBlob::TruncateFile : Failed to get block list of file %s [%s]", name, err.Error())
+		return err
+	}
+
+	if committedBlockList == nil {
+		// This file is not split into blocks yet
+		// Read full file and then proceed based on new size
+		data, err := bb.readFileData(name, size, attr.Size)
 		if err != nil {
-			log.Err("BlockBlob::TruncateFile : Failed to write from buffer file %s", name, err.Error())
+			log.Err("BlockBlob::TruncateFile : Failed to read file %s [%s]", name, err.Error())
 			return err
 		}
-	} else {
-		bol, err := bb.GetFileBlockOffsets(name)
+
+		err = bb.convertFileToBlocks(name, size, blockSize, data)
 		if err != nil {
-			log.Err("BlockBlob::TruncateFile : Failed to get block list of file %s [%s]", name, err.Error())
-			return err
+			log.Err("BlockBlob::TruncateFile : Failed to expand file %s [%s]", name, err.Error())
 		}
-		if bol.SmallFile() {
-			data, err := bb.HandleSmallFile(name, size, attr.Size)
-			if err != nil {
-				log.Err("BlockBlob::TruncateFile : Failed to read small file %s", name, err.Error())
-				return err
-			}
-			err = bb.WriteFromBuffer(name, nil, data)
-			if err != nil {
-				log.Err("BlockBlob::TruncateFile : Failed to write from buffer file %s", name, err.Error())
-				return err
-			}
-		} else {
-			if size < attr.Size {
-				bol = bb.removeBlocks(bol, size, name)
-			} else if size > attr.Size {
-				_, err = bb.createNewBlocks(bol, bol.BlockList[len(bol.BlockList)-1].EndIndex, size-attr.Size)
-				if err != nil {
-					log.Err("BlockBlob::TruncateFile : Failed to create new blocks for file %s", name, err.Error())
-					return err
-				}
+		return err
+	}
+
+	// This file already has the blocks created we just need to truncate the blocks
+	return bb.updateBlocks(name, size, blockSize, committedBlockList)
+}
+
+func (bb *BlockBlob) readFileData(name string, truncateSize int64, fileSize int64) ([]byte, error) {
+	// If truncate size is smaller then file size then read only data upto truncate size
+	// If truncate size is bigger then file size then read entire file data
+	size := int64(math.Min((float64)(truncateSize), (float64)(fileSize)))
+
+	var data = make([]byte, size)
+
+	err := bb.ReadInBuffer(name, 0, size, data)
+	if err != nil {
+		log.Err("BlockBlob::readFileData : Failed to read small file %s", name, err.Error())
+	}
+
+	return data, err
+}
+
+func (bb *BlockBlob) createSparseFile(name string, size int64, blockSize int64, blkList []string, idLen int64) error {
```

Comment:
It is better to have 2 functions originating from truncate function which are 
1. Expand
2. Shrink
This will make the code more readable